lib_eio_posix/eio_posix_stubs.c

```diff
@@ -0,0 +1,178 @@
+#else
+    arc4random_buf(buf, len);
+    ret = len;
+#endif
+    caml_leave_blocking_section();
+  } while (ret == -1 && errno == EINTR);
+  if (ret == -1) uerror("getrandom", Nothing);
+  CAMLreturn(Val_long(ret));
+}
+
+/* Fill [iov] with pointers to the cstructs in the array [v_bufs]. */
+static void fill_iov(struct iovec *iov, value v_bufs) {
+  int n_bufs = Wosize_val(v_bufs);
+  for (int i = 0; i < n_bufs; i++) {
+    value v_cs = Field(v_bufs, i);
+    value v_ba = Field(v_cs, 0);
+    value v_off = Field(v_cs, 1);
+    value v_len = Field(v_cs, 2);
+    iov[i].iov_base = Caml_ba_data_val(v_ba) + Long_val(v_off);
+    iov[i].iov_len = Long_val(v_len);
+  }
+}
+
+CAMLprim value caml_eio_posix_readv(value v_fd, value v_bufs) {
+  CAMLparam1(v_bufs);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = readv(Int_val(v_fd), iov, n_bufs);
+  if (r < 0) uerror("readv", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_writev(value v_fd, value v_bufs) {
+  CAMLparam1(v_bufs);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = writev(Int_val(v_fd), iov, n_bufs);
+  if (r < 0) uerror("writev", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_preadv(value v_fd, value v_bufs, value v_offset) {
+  CAMLparam2(v_bufs, v_offset);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = preadv(Int_val(v_fd), iov, n_bufs, Int63_val(v_offset));
+  if (r < 0) uerror("preadv", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_pwritev(value v_fd, value v_bufs, value v_offset) {
+  CAMLparam2(v_bufs, v_offset);
+  ssize_t r;
+  int n_bufs = Wosize_val(v_bufs);
+  struct iovec iov[n_bufs];
+
+  fill_iov(iov, v_bufs);
+
+  r = pwritev(Int_val(v_fd), iov, n_bufs, Int63_val(v_offset));
+  if (r < 0) uerror("pwritev", Nothing);
+
+  CAMLreturn(Val_long(r));
+}
+
+CAMLprim value caml_eio_posix_openat(value v_dirfd, value v_pathname, value v_flags, value v_mode) {
+  CAMLparam1(v_pathname);
+  char* pathname;
+  int r;
+
+  caml_unix_check_path(v_pathname, "openat");
+  pathname = caml_stat_strdup(String_val(v_pathname));
+
+  caml_enter_blocking_section();
+  r = openat(Int_val(v_dirfd), pathname, Int_val(v_flags), Int_val(v_mode));
+  caml_leave_blocking_section();
+
+  caml_stat_free(pathname);
+  if (r < 0) uerror("openat", v_pathname);
+  CAMLreturn(Val_int(r));
+}
+
+CAMLprim value caml_eio_posix_mkdirat(value v_fd, value v_path, value v_perm) {
+  CAMLparam1(v_path);
+  char *path;
+  int ret;
+  caml_unix_check_path(v_path, "mkdirat");
+  path = caml_stat_strdup(String_val(v_path));
+  caml_enter_blocking_section();
+  ret = mkdirat(Int_val(v_fd), path, Int_val(v_perm));
+  caml_leave_blocking_section();
+  caml_stat_free(path);
+  if (ret == -1) uerror("mkdirat", v_path);
```

Comment:
are you sure that *free(2)* won't clobber errno? It was a bit [muddy in the past](https://sourceware.org/bugzilla/show_bug.cgi?id=17924) but might be the case now.  To be safe, should save errno right after the function call before any other functions.  But calling free might be safe nowadays...

Reply:
Interesting. https://github.com/ocaml/ocaml/blob/trunk/otherlibs/unix/mkdir.c seems to assume errno won't change.

Reply:
:fearful: I had no idea glibc could do that, that's quite surprising.

Reply:
malloc and free are pretty complex beasts on Linux, but the errno-setting happens in pretty specific cases (usually when using the mmap mode).  Irritatingly hard to predict, but triggers a more often in virtualised platforms where there are a lot of VMA games going in with the virtual memory subsystem.

Reply:
I've now wrapped the free function as `caml_stat_free_preserving_errno` to avoid that.